numpyro/infer/hmc_util.py

```diff
@@ -350,6 +379,65 @@ def build_adaptation_schedule(num_steps):
     return adaptation_schedule
 
 
+def _initialize_mass_matrix(z, inverse_mass_matrix, dense_mass):
+    if isinstance(dense_mass, list):
+        inverse_mass_matrix = {} if inverse_mass_matrix is None else inverse_mass_matrix
+        mass_matrix_sqrt = {}
+        mass_matrix_sqrt_inv = {}
+        for site_names in dense_mass:
+            inverse_mm = inverse_mass_matrix.get(site_names)
+            z_block = tuple(z[k] for k in site_names)
+            inverse_mm, mm_sqrt, mm_sqrt_inv = _initialize_mass_matrix(z_block, inverse_mm, True)
+            inverse_mass_matrix[site_names] = inverse_mm
+            mass_matrix_sqrt[site_names] = mm_sqrt
+            mass_matrix_sqrt_inv[site_names] = mm_sqrt_inv
+        # NB: this branch only happens when users want to use block diagonal
+        # inverse_mass_matrix, for example, {("a",): jnp.ones(3), ("b",): jnp.ones(3)}.
+        for site_names, inverse_mm in inverse_mass_matrix.items():
```

Comment:
hmm this part is kind of tricky you may need to walk me through this...

Reply:
I guess this is resolved per our discussion.